File: GlazerAI/CLI/ClaudeRunner.swift
// ClaudeRunner.swift
// GlazerAI
//
// Actor that invokes the `claude -p` CLI and captures its output.
// Runs the process asynchronously with a configurable timeout.

import Foundation

// MARK: - Errors

/// Errors produced by ``ClaudeRunner``.
enum ClaudeError: LocalizedError {
    /// The `claude` CLI binary was not found on PATH.
    case notFound
    /// The process exceeded the allowed timeout.
    case timeout
    /// The process exited with a non-zero status code.
    case executionFailed(stderr: String)

    var errorDescription: String? {
        switch self {
        case .notFound:
            return "The Claude CLI was not found. "
                + "Please install it from https://claude.ai/download"
        case .timeout:
            return "The Claude CLI did not respond within 120 seconds."
        case .executionFailed(let stderr):
            let detail = stderr.isEmpty ? "Unknown error" : stderr
            return "Claude CLI failed: \(detail)"
        }
    }
}

// MARK: - Output Envelope

private struct ClaudeUsage: Decodable {
    let inputTokens: Int
    let outputTokens: Int

    enum CodingKeys: String, CodingKey {
        case inputTokens  = "input_tokens"
        case outputTokens = "output_tokens"
    }
}

/// The JSON envelope that `claude -p --output-format json` wraps its response in.
private struct ClaudeOutputEnvelope: Decodable {
    let isError: Bool
    let result: String
    let usage: ClaudeUsage?

    enum CodingKeys: String, CodingKey {
        case isError = "is_error"
        case result
        case usage
    }
}

// MARK: - Runner

/// Invokes `claude -p --output-format json "<prompt>"` and returns the inner result text.
actor ClaudeRunner {

    /// Timeout in seconds for the claude process.
    private let timeoutSeconds: TimeInterval

    /// Creates a runner with the given timeout.
    /// - Parameter timeoutSeconds: Maximum time to wait for the process. Defaults to 60.
    init(timeoutSeconds: TimeInterval = 120) {
        self.timeoutSeconds = timeoutSeconds
    }

    /// Runs the claude CLI with the given prompt and returns the result text.
    ///
    /// Uses `--output-format json` so the process output is a structured envelope,
    /// giving reliable `is_error` detection and clean extraction of the response text.
    ///
    /// - Parameter prompt: The assembled research prompt to send.
    /// - Returns: The `result` field from the JSON envelope.
    /// - Throws: ``ClaudeError`` on failure.
    func run(prompt: String) async throws -> String {
        guard let claudePath = await CLIEnvironment.shared.claudePath else {
            throw ClaudeError.notFound
        }

        let raw = try await withCheckedThrowingContinuation { continuation in
            executeProcess(
                claudePath: claudePath,
                prompt: prompt,
                continuation: continuation
            )
        }

        return try extractResult(from: raw)
    }

    // MARK: - Private

    private func executeProcess(
        claudePath: String,
        prompt: String,
        continuation: CheckedContinuation<String, Error>
    ) {
        let process = Process()
        let stdinPipe = Pipe()
        let stdoutPipe = Pipe()
        let stderrPipe = Pipe()

        process.executableURL = URL(fileURLWithPath: "/bin/zsh")
        process.arguments = ["-l", "-c",
            "\(claudePath) -p --output-format json --allowedTools web_search"
        ]
        process.standardInput = stdinPipe
        process.standardOutput = stdoutPipe
        process.standardError = stderrPipe

        let timeoutItem = DispatchWorkItem { [weak process] in
            process?.terminate()
        }
        DispatchQueue.global().asyncAfter(
            deadline: .now() + timeoutSeconds,
            execute: timeoutItem
        )

        do {
            try process.run()

            // Write prompt to stdin then close so the process sees EOF.
            if let data = prompt.data(using: .utf8) {
                stdinPipe.fileHandleForWriting.write(data)
            }
            stdinPipe.fileHandleForWriting.closeFile()

            process.waitUntilExit()
            timeoutItem.cancel()

            let result = readRawOutput(process: process, stdout: stdoutPipe, stderr: stderrPipe)
            continuation.resume(with: result)
        } catch {
            timeoutItem.cancel()
            continuation.resume(throwing: ClaudeError.executionFailed(
                stderr: error.localizedDescription
            ))
        }
    }

    private func readRawOutput(
        process: Process,
        stdout: Pipe,
        stderr: Pipe
    ) -> Result<String, Error> {
        let stdoutData = stdout.fileHandleForReading.readDataToEndOfFile()
        let stderrData = stderr.fileHandleForReading.readDataToEndOfFile()
        let stdoutStr = String(data: stdoutData, encoding: .utf8) ?? ""
        let stderrStr = String(data: stderrData, encoding: .utf8) ?? ""

        if process.terminationReason == .uncaughtSignal {
            return .failure(ClaudeError.timeout)
        } else if process.terminationStatus != 0 {
            return .failure(ClaudeError.executionFailed(stderr: stderrStr))
        } else {
            return .success(stdoutStr)
        }
    }

    /// Decodes the JSON envelope and returns `result`, or throws if `is_error` is true.
    private func extractResult(from envelopeJSON: String) throws -> String {
        let data = Data(envelopeJSON.utf8)
        let envelope = (try? JSONDecoder().decode(ClaudeOutputEnvelope.self, from: data))
        guard let env = envelope else {
            // Fallback: envelope not parseable, return raw text as-is
            return envelopeJSON.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        if env.isError {
            throw ClaudeError.executionFailed(stderr: env.result)
        }
        if let usage = env.usage {
            SessionUsage.shared.record(
                inputTokens: usage.inputTokens,
                outputTokens: usage.outputTokens
            )
        }
        return env.result
    }

}

File: electron-app/src/main/services/claude-runner.ts
import { execFile } from 'child_process';
import { promisify } from 'util';
import { resolve as resolvePath } from 'path';

const execFileAsync = promisify(execFile);

/** JSON envelope that `claude -p --output-format json` returns. */
interface ClaudeOutputEnvelope {
  is_error: boolean;
  result: string;
  usage?: {
    input_tokens: number;
    output_tokens: number;
  };
}

/** Timeout for the claude process in milliseconds. */
const TIMEOUT_MS = 120_000;

/**
 * Resolves the path to the `claude` CLI binary.
 * Checks common install locations, then falls back to PATH.
 */
export function findClaudePath(): string | null {
  const candidates = [
    resolvePath(process.env.HOME || '~', '.local', 'bin', 'claude'),
    '/usr/local/bin/claude',
    '/opt/homebrew/bin/claude',
  ];

  // On Windows, check AppData
  if (process.platform === 'win32') {
    const appData = process.env.LOCALAPPDATA || '';
    if (appData) {
      candidates.push(resolvePath(appData, 'Programs', 'claude', 'claude.exe'));
    }
    candidates.push('claude.exe');
  }

  // Try each candidate
  const { execFileSync } = require('child_process');
  for (const candidate of candidates) {
    try {
      execFileSync(candidate, ['--version'], { timeout: 5000, stdio: 'pipe' });
      return candidate;
    } catch {
      // Not found at this path, try next.
    }
  }

  // Fall back to `which` / `where`
  try {
    const cmd = process.platform === 'win32' ? 'where' : 'which';
    const result = execFileSync(cmd, ['claude'], { timeout: 5000, stdio: 'pipe' });
    const path = result.toString().trim().split('\n')[0];
    if (path) return path;
  } catch {
    // Not on PATH.
  }

  return null;
}

/**
 * Runs `claude -p --output-format json --allowedTools web_search` with
 * the prompt piped via stdin. Returns the result text from the JSON envelope.
 *
 * Mirrors the Swift ClaudeRunner exactly.
 */
export async function runClaude(prompt: string): Promise<string> {
  const claudePath = findClaudePath();
  if (!claudePath) {
    throw new Error(
      'Claude CLI not found. Install it from https://claude.ai/download'
    );
  }

  // Use a shell to invoke claude so it picks up the user's login environment.
  const shell = process.platform === 'win32' ? 'cmd.exe' : '/bin/zsh';
  const shellArgs =
    process.platform === 'win32'
      ? ['/c', `"${claudePath}" -p --output-format json --allowedTools web_search`]
      : ['-l', '-c', `${claudePath} -p --output-format json --allowedTools web_search`];

  return new Promise<string>((resolve, reject) => {
    const { spawn } = require('child_process');
    const proc = spawn(shell, shellArgs, {
      timeout: TIMEOUT_MS,
      stdio: ['pipe', 'pipe', 'pipe'],
      env: { ...process.env },
    });

    let stdout = '';
    let stderr = '';

    proc.stdout.on('data', (data: Buffer) => {
      stdout += data.toString();
    });

    proc.stderr.on('data', (data: Buffer) => {
      stderr += data.toString();
    });

    proc.on('error', (err: Error) => {
      reject(new Error(`Failed to start Claude CLI: ${err.message}`));
    });

    proc.on('close', (code: number | null, signal: string | null) => {
      if (signal === 'SIGTERM' || signal === 'SIGKILL') {
        reject(new Error('Claude CLI did not respond within 120 seconds.'));
        return;
      }

      if (code !== 0) {
        reject(
          new Error(`Claude CLI failed: ${stderr || 'Unknown error'}`)
        );
        return;
      }

      try {
        const result = extractResult(stdout);
        resolve(result);
      } catch (err) {
        reject(err);
      }
    });

    // Write prompt to stdin, then close to signal EOF.
    proc.stdin.write(prompt);
    proc.stdin.end();
  });
}

/**
 * Decodes the JSON envelope and returns the result field.
 * Falls back to raw output if the envelope is not parseable.
 */
function extractResult(envelopeJSON: string): string {
  const trimmed = envelopeJSON.trim();

  let envelope: ClaudeOutputEnvelope;
  try {
    envelope = JSON.parse(trimmed);
  } catch {
    // Envelope not parseable, return raw text.
    return trimmed;
  }

  if (envelope.is_error) {
    throw new Error(`Claude CLI error: ${envelope.result}`);
  }

  return envelope.result;
}
